File: cordova-plugin-globeconnect/src/ios/Sms.swift
//
//  Sms.swift
//
//  Created by Rico Maglayon on 08/12/2016.
//

import Foundation

@objc(Sms) class Sms : CDVPlugin {
    var smsSenderAddress: String?
    var smsAccessToken: String?
    var smsClientCorrelator: String?
    var smsReceiverAddress: String?
    var smsMessage: String?

    @objc(setSenderAddress:) func setSenderAddress(command: CDVInvokedUrlCommand) -> Sms {
        let argument = command.arguments[0] as? String ?? ""
        self.smsSenderAddress = argument

        return self
    }

    @objc(setAccessToken:) func setAccessToken(command: CDVInvokedUrlCommand) -> Sms {
        let argument = command.arguments[0] as? String ?? ""
        self.smsAccessToken = argument

        return self
    }

    @objc(setClientCorrelator:) func setClientCorrelator(command: CDVInvokedUrlCommand) -> Sms {
        let argument = command.arguments[0] as? String ?? ""
        self.smsClientCorrelator = argument

        return self
    }

    @objc(setReceiverAddress:) func setReceiverAddress(command: CDVInvokedUrlCommand) -> Sms {
        let argument = command.arguments[0] as? String ?? ""
        self.smsReceiverAddress = argument

        return self
    }

    @objc(setMessage:) func setMessage(command: CDVInvokedUrlCommand) -> Sms {
        let argument = command.arguments[0] as? String ?? ""
        self.smsMessage = argument

        return self
    }

    @objc(sendMessage:) func sendMessage(command: CDVInvokedUrlCommand) -> Sms {
        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )

        // set the url
        let sendMessageUrl = "https://devapi.globelabs.com.ph/smsmessaging/v1/outbound/"+self.smsSenderAddress!+"/requests?access_token="+self.smsAccessToken!

        // set the header/s
        var headers = Dictionary<String, String>()
        headers["Content-Type"] = "application/json; charset=utf-8"

        // fix the address
        let address = "tel:" + self.smsReceiverAddress!

        // fix the sender address
        let sender = "tel:" + self.smsSenderAddress!

        // prepare the payload
        let data: [String : [String: Any]] = [
            "outboundSMSMessageRequest" : [
                "senderAddress"           : sender,
                "address"                 : [address],
                "outboundSMSTextMessage"  : [
                    "message" : self.smsMessage!
                ]
            ]
        ]

        // we need to convert first the payload to JSON
        do {
            // convert it!
            let jsonData = try JSONSerialization.data(
                withJSONObject: data,
                options: JSONSerialization.WritingOptions.prettyPrinted
            )

            // it is now in json so we need it to be a string so we can send it
            if let jsonPayload = String(data: jsonData, encoding: String.Encoding.utf8) {
                HTTPRequest(url: sendMessageUrl, headers: headers)
                    .post(
                        parameters: jsonPayload,
                        success: { data, _ in
                            DispatchQueue.global(qos: .utility).async {
                                DispatchQueue.main.async {
                                    pluginResult = CDVPluginResult(
                                        status: CDVCommandStatus_OK,
                                        messageAs: String(data: data, encoding: .utf8)
                                    )

                                    self.commandDelegate!.send(
                                        pluginResult,
                                        callbackId: command.callbackId
                                    )
                                }
                            }
                    }, failure: { error, data in
                        pluginResult = CDVPluginResult(
                            status: CDVCommandStatus_ERROR,
                            messageAs: String(data: data, encoding: .utf8)
                        )

                        self.commandDelegate!.send(
                            pluginResult,
                            callbackId: command.callbackId
                        )
                    })
            }
        } catch let error as NSError {
            // oops, error in converting it to JSON
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: error.localizedDescription
            )

            self.commandDelegate!.send(
                pluginResult,
                callbackId: command.callbackId
            )
        }

        return self
    }
}

File: cordova-plugin-globeconnect/www/Sms.js
var exec = require('cordova/exec');

// Sms
var Sms = function(senderAddress, accessToken) {
    // short hand initialization
    if(!(this instanceof Sms)) {
        return new Sms(senderAddress, accessToken);
    }

    // class name
    this.class = 'Sms';

    // sender address set?
    if(senderAddress) {
        exec(null, null, this.class, 'setSenderAddress', [senderAddress]);
    }

    // access token set?
    if(accessToken) {
        exec(null, null, this.class, 'setAccessToken', [accessToken]);
    }

    /**
     * Set sender address.
     *
     * @param  string
     * @return this
     */
    this.setSenderAddress = function(senderAddress) {
        exec(null, null, this.class, 'setSenderAddress', [senderAddress]);

        return this;
    };

    /**
     * Set access token.
     *
     * @param  string
     * @return this
     */
    this.setAccessToken = function(accessToken) {
        exec(null, null, this.class, 'setAccessToken', [accessToken]);

        return this;
    };

    /**
     * Set client correlator.
     *
     * @param  string
     * @return this
     */
    this.setClientCorrelator = function(clientCorrelator) {
        exec(null, null, this.class, 'setClientCorrelator', [clientCorrelator]);

        return this;
    };

    /**
     * Set receiver address.
     *
     * @param  string
     * @return this
     */
    this.setReceiverAddress = function(receiverAddress) {
        exec(null, null, this.class, 'setReceiverAddress', [receiverAddress]);

        return this;
    };

    /**
     * Set message.
     *
     * @param  string
     * @return this
     */
    this.setMessage = function(message) {
        exec(null, null, this.class, 'setMessage', [message]);

        return this;
    };

    /**
     * Send message.
     *
     * @param  function
     * @param  function
     * @return this
     */
    this.sendMessage = function(successCallback, errorCallback) {
        var callback = function(data) {
            try {
                data = JSON.parse(data);
            } catch(e) {};

            successCallback.call(this, data);
        };

        exec(callback, errorCallback, this.class, 'sendMessage', []);

        return this;
    };
};

module.exports = Sms;
